### PR TITLE
refactor: Replace vector with array in C code

### PR DESCRIPTION
**Pull request by krahets**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference code.
- [x] I've checked the code (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by krahets** [comment]
@lhxsm @Gonglja Please help check the correctness of the code, thx!

File: codes/c/chapter_backtracking/preorder_traversal_ii_compact.c [hunk 1]
```diff
@@ -6,62 +6,50 @@
 
 #include "../utils/common.h"
 
+// 假设结果长度不超过 100
+#define MAX_SIZE 100
+
+TreeNode *path[MAX_SIZE];
+TreeNode *res[MAX_SIZE][MAX_SIZE];
+int pathSize = 0, resSize = 0;
+
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/preorder_traversal_ii_compact.c:15]: Define an array to represent the length of each path.
But it's not a big problem when you see the traversal results later on.
`for (int j = 0; res[i][j] != NULL; ++j) {`

File: codes/c/chapter_backtracking/preorder_traversal_iii_compact.c [hunk 2]
```diff
@@ -6,64 +6,53 @@
...
+        }
+        printArray(vals, size);
+    }
 
-    // 释放内存
-    delVector(path);
-    delVector(res);
     return 0;
-}
+}
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/preorder_traversal_iii_compact.c:58]: Use `freeMemoryTree` to free the memory allocated by `arrayToTree`.

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 3]
```diff
@@ -6,96 +6,84 @@
...
         }
-        printVector(vals, printFunc);
+        printArray(vals, size);
     }
 
-    // 释放内存
-    delVector(state);
-    delVector(choices);
-    delVector(res);
     return 0;
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:88]: Above as

File: codes/c/chapter_backtracking/preorder_traversal_iii_template.c [hunk 4]
```diff
@@ -6,96 +6,84 @@
 
 #include "../utils/common.h"
 
+// 假设结果长度不超过 100
+#define MAX_SIZE 100
+
+TreeNode *path[MAX_SIZE];
+TreeNode *res[MAX_SIZE][MAX_SIZE];
+int pathSize = 0, resSize = 0;
+
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/preorder_traversal_iii_template.c:15]: Same comments as in the previous document(`codes/c/chapter_backtracking/preorder_traversal_ii_compact.c`)

File: codes/c/chapter_backtracking/subset_sum_i.c [hunk 5]
```diff
@@ -6,73 +6,72 @@
...
 
 /* Driver Code */
 int main() {
     int nums[] = {3, 4, 5};
-    vector *iNums = newVector();
-    for (int i = 0; i < sizeof(nums) / sizeof(nums[0]); i++) {
-        vectorPushback(iNums, &nums[i], sizeof(int));
-    }
-
+    int numsSize = 3;
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/subset_sum_i.c:63]: `sizeof(nums)/sizeof(nums[0])`

File: codes/c/chapter_backtracking/subset_sum_i.c [hunk 6]
```diff
@@ -6,73 +6,72 @@
...
 
+#define MAX_SIZE 100
+
+// 状态（子集）
+int state[MAX_SIZE];
+int stateSize = 0;
+
+// 结果列表（子集列表）
+int res[MAX_SIZE][MAX_SIZE];
+int resColSizes[MAX_SIZE];
```
Comment (Gonglja) [inline codes/c/chapter_backtracking/subset_sum_i.c:17]: Good. That's what I meant.

File: codes/c/chapter_tree/array_binary_tree.c [hunk 7]
```diff
@@ -6,29 +6,31 @@
...
 
-/* 构造函数 */
-ArrayBinaryTree *newArrayBinaryTree(vector *arr) {
-    ArrayBinaryTree *newABT = malloc(sizeof(ArrayBinaryTree));
-    newABT->tree = arr;
-    return newABT;
+/* 构造方法 */
+ArrayBinaryTree *createArrayBinaryTree(int *arr, int size) {
+    ArrayBinaryTree *abt = (ArrayBinaryTree *)malloc(sizeof(ArrayBinaryTree));
+    abt->tree = arr;
```
Comment (Gonglja) [inline codes/c/chapter_tree/array_binary_tree.c:18]: It is recommended to reallocate a section of space, copy the external data to the internal, and release it after subsequent use.
```c
abt->tree = malloc (sizeof(int) * size);
memcpy(abt->tree, arr, size);
```

File: codes/c/chapter_tree/array_binary_tree.c [hunk 8]
```diff
@@ -46,114 +48,113 @@ int parent(int i) {
...
+    printArray(res, returnSize);
+    free(res);
+
+    // 释放内存
+    free(abt);
 
     return 0;
-}
\ No newline at end of file
+}
```
Comment (Gonglja) [inline codes/c/chapter_tree/array_binary_tree.c:160]: Use `freeMemoryTree` to free the memory allocated by `arrayToTree`.

**Review by Gonglja — COMMENTED** [review]
Hi @krahets , I've review it.
Let's see if there are any different opinions, feel free to discuss.

File: codes/c/chapter_tree/array_binary_tree.c (hunk 7 above)
Comment (krahets) [reply codes/c/chapter_tree/array_binary_tree.c:18]: The second line should be:

```c
memcpy(abt->tree, arr, sizeof(int) * size);
```

**Comment by Gonglja** [comment]
Yes, negligence

发自我的iPhone

在 2023年10月27日，23:21，Yudong Jin ***@***.***> 写道：

﻿

@krahets commented on this pull request.

________________________________

In codes/c/chapter_tree/array_binary_tree.c<https://github.com/krahets/hello-algo/pull/894#discussion_r1374728517>:

>  } ArrayBinaryTree;

-/* 构造函数 */
-ArrayBinaryTree *newArrayBinaryTree(vector *arr) {
-    ArrayBinaryTree *newABT = malloc(sizeof(ArrayBinaryTree));
-    newABT->tree = arr;
-    return newABT;
+/* 构造方法 */
+ArrayBinaryTree *createArrayBinaryTree(int *arr, int size) {
+    ArrayBinaryTree *abt = (ArrayBinaryTree *)malloc(sizeof(ArrayBinaryTree));
+    abt->tree = arr;


The second line should be:

memcpy(abt->tree, arr, sizeof(int) * size);

**Comment by krahets** [comment]
Thanks for the review @Gonglja !

**Pull request merged**